blocks/api/categories.js

```diff
@@ -20,10 +23,111 @@ const categories = [
 ];
 
 /**
- * Returns all the block categories
+ * Returns all the block categories.
  *
- * @return {Array} Block categories
+ * @return {Array} categories
  */
 export function getCategories() {
 	return categories;
 }
+
+/**
+ * Register a new block Category.
+ *
+ * @param {Array} category e.g {slug: 'custom', title: __('Custom Blocks')}
+ *
+ * @return {Array} categories
+ */
+export function registerCategory( category ) {
+	if ( ! category ) {
+		console.error(
+			'The block Category must be defined'
+		);
+		return;
+	}
+	if ( ! category.slug ) {
+		console.error(
+			'The block Category slug must be defined'
+		);
+		return;
+	}
+	if ( ! /^[a-z0-9-]+$/.test( category.slug ) ) {
+		console.error(
+			'The block Category slug must not contain characters which are invalid for urls'
+		);
+		return;
+	}
+	if ( categories.find( x => x.slug === category.slug ) ) {
```

Comment:
note that `Array.prototype.some()` here might have a more semantic match here with what we're checking: is there any item in this collection which matches the given predicate function?